nativelib/src/main/resources/gc/immix/Heap.h

```diff
@@ -0,0 +1,47 @@
+#ifndef IMMIX_HEAP_H
+#define IMMIX_HEAP_H
+
+#include "GCTypes.h"
+#include "Allocator.h"
+#include "LargeAllocator.h"
+#include "datastructures/Stack.h"
+
+typedef struct {
+    word_t *heapStart;
+    word_t *heapEnd;
+    size_t smallHeapSize;
+    word_t *largeHeapStart;
+    word_t *largeHeapEnd;
+    word_t *largeHeapSize;
+    Allocator *allocator;
+    LargeAllocator *largeAllocator;
+} Heap;
+
+static inline bool heap_isWordInLargeHeap(Heap *heap, word_t *word) {
+    return word != NULL && word >= heap->largeHeapStart &&
+           word < heap->largeHeapEnd;
+}
+
+static inline bool heap_isWordInSmallHeap(Heap *heap, word_t *word) {
+    return word != NULL && word >= heap->heapStart && word < heap->heapEnd;
+}
+
+static inline bool heap_isWordInHeap(Heap *heap, word_t *word) {
+    return heap_isWordInSmallHeap(heap, word) ||
+           heap_isWordInLargeHeap(heap, word);
+}
```

Comment:
Yes and no. We can do it but it will not help. The problem being that we can map memory only once for both heaps and grow up and down. But we can have a value from the stack pointing to "unallocated heap" meaning part of memory that has been mapped but where the heaps have not grown to yet.